doc/designs/app-passwords.md

```diff
@@ -0,0 +1,312 @@
+particular service) to use the new one.
+
+This enhancement therefore brings support for app passwords (sometimes also
+called application-specific passwords), a form of multiple passwords for a single
+user. Only this would, if app passwords are each used only on one device, increase
+convenience in the case that the user can no longer trust a password used on one
+of his devices. Namely, he can revoke or change only that password, without
+needing to revoke or change his other passwords, including the primary password.
+What is more, app passwords will generally not be usable for account management.
+
+As a further possible future enhancement, this design page also outlines a
+proposal to add support for use of app passwords with FreeIPA's Kerberos
+functionality, what would enable to use the HBAC (host-based access control) to
+allow specifying what app passwords can be used to log into a particular system
+or service. The use of HBAC with app passwords would add another security
+benefit on top of those mentioned above - the intruder's access to the user's
+account might be denied not only to the account management, but also to at least
+some of the systems and services that it is used for.
+
+Resolves: https://pagure.io/freeipa/issue/4510
+
+How Google does it: https://support.google.com/accounts/answer/185833?hl=en
+
+## Use cases
+
+### Use with password-only authentication
+
+The user can set multiple passwords per account. Each password can be then used
+for authenticating on only one device. Various passwords can as well be used to
+authenticate to various services or their sets. These divisions can even be
+combined, thus - in the most extreme case - having one password for
+authenticating to one specific service on one particular device. Granularity of
+such divisions will be fully up to user's discretion.
+
+### Use as a part of 2FA (OTP)
+
+When 2FA will be enabled, it will work (and be required) only for the user's
+primary password, not for any of the app passwords, as that would cause too much
+trouble.
+
+### Use as a fallback mechanism when 2FA is activated, but the client app does not support it
+
+As 2FA will not be applied to app passwords, the user can use an app password to
+log into a service that does not support 2FA.
+
+## Design
+
+### Multiple passwords for a single user and related changes
+
+App passwords will be stored in a new LDAP subtree in the form
+```
+dn: uid=user-1,cn=myapp,cn=apps,cn=accounts,$SUFFIX
+uid: user-1
+displayName: My eMail app password for tablet
+appname: myapp
+entryUUID: XXXXXXXXX-XXXXXXXXXX-XXXXXXXXXX
+userPassword: *********************
+managedby: uid=user,cn=users,cn=accounts,$SUFFIX
+
+dn: uid=user-2,cn=myapp,cn=apps,cn=accounts,$SUFFIX
+uid: user-2
+displayName: My eMail app password for smartphone
+appname: myapp
+entryUUID: XXXXXXXXX-XXXXXXXXXX-XXXXXXXXXX
+userPassword: *********************
+managedby: uid=user,cn=users,cn=accounts,$SUFFIX
```

Comment:
managedby is multi-valued. Have you thought about how delegation might work? Are the ACI's going to center around managedby? This would allow another delegated entry to change the password.

Reply:
I did not think about delegation yet. `managedby` will be set automatically to the DN of the object representing the user (for example, `uid=user,cn=users,cn=accounts,$SUFFIX`), to which the password belongs. It will then be necessary to maintain referential integrity, if this DN of the user changes - I did not think through details of this yet. Is it even possible to rename or move the object representing the user? If not, then the value of the `managedby` attribute will remain the same from the creation of the app password object until the deletion of the app password object. The drafted ACIs give add, search and delete permissions to the `managedby` attribute, but not the write permission, so if this is the case, I assume I could leave it like this, or not?

Reply:
As for delegation, for it to work, there must be an ACI allowing it, right? Is this actually a problem? I don't know how exactly this feature is used, I only know about this feature from [here](https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/7/html/linux_domain_identity_authentication_and_policy_guide/server-access-controls#types-of-acls).

Reply:
Reading back in the comments I see this is already somewhat covered, memberof vs managedby. You are correct that an ACI would be needed to write to managedby but typically admins can write to anything. But managedby is use explicitly elsewhere to allow delegation. For example in services to allow other hosts to obtain certificates for a service.

The IPA vault, for example, uses the owner attribute to do the linkage. That is probably clearer.

But then again, you propose future use-cases using services and hosts and managedby would be completely appropriate in that context.